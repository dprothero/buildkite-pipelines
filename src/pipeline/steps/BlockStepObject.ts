import { Branches, Conditional, Dependencies, Key, Prompt } from "./step-objects";

export interface BlockStepObject extends Key, Dependencies, Conditional, Branches, Prompt {
  block: string;
  blocked_state?: 'passed' | 'failed' | 'running'
}