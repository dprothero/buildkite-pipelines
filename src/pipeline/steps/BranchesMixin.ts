
export interface BranchesMixinMethods {
  branches(branches: string[]): this;
  addBranch(branch: string): this
}

export interface BranchesMixinBuilder extends BranchesMixinMethods {
  build(): {branches?: string[]} 
}

export class BranchesMixin {
  static builder(): BranchesMixinBuilder {
    let _branches: string[] | undefined
    return {
      branches(branches) {
        _branches = branches;
        return this
      },
      addBranch(branch) {
        // TODO: set to array when branches is undefiend
        
        _branches?.push(branch);
        return this
      },
      build() {
        return {
          branches: _branches
        }
      }
    }
  }
}
