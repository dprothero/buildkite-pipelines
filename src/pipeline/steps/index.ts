export * from "./BlockStepObject";
// export * from  './BlockStep';
export * from "./CommandStep";
export * from "./CommandStepObject";
export * from "./StepBuilder";
export * from "./GroupStepObject";
export * from "./GroupStep";
export * from "./InputStepObject";
export * from "./StepBuilder";
export * from "./StepObject";
// export * from  './TriggerStep';
export * from "./TriggerStepObject";
export * from "./WaitStep";
export * from "./WaitStepObject";
export * from "./plugins";
