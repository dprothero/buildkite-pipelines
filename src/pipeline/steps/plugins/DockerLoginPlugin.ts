import { PluginObject } from "./PluginObject";
import { PluginBuilder } from "./PluginBuilder";

export interface DockerLoginPluginObject {
  server?: string;
  username: string;
  'password-env': string
}

export class DockerLoginPlugin implements PluginBuilder {
  #server: string | undefined;
  #username: string | undefined;
  #passwordEnv: string | undefined;
  #retries: number | undefined;

  server(): string | undefined;
  server(server: string | undefined): this;
  server(server?: string | undefined): string | undefined | this {
    if (server) {
      this.#server = server
      return this
    } else {
      return this.#server;
    }
  }

  username(): string;
  username(username: string): this;
  username(username?: string): string | this {
    if (username) {
      this.#username = username
      return this
    } else {
      if (!this.#username) {
        throw new Error('username not specified')
      }
      return this.#username;
    }
  }

  passwordEnv(): string;
  passwordEnv(passwordEnv: string): this;
  passwordEnv(passwordEnv?: string): string | this {
    if (passwordEnv) {
      this.#passwordEnv = passwordEnv
      return this
    } else {
      if (!this.#passwordEnv) {
        throw new Error('passwordEnv not specified')
      }
      return this.#passwordEnv;
    }
  }

  public build(): PluginObject<DockerLoginPluginObject> {
    return {
      'docker-login#v2.1.0': {
        server: this.server(),
        username: this.username(),
        "password-env": this.passwordEnv(),
      }
    }
  }
}