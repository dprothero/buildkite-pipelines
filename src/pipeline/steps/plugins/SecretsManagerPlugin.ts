import { PluginObject } from "./PluginObject";
import { PluginBuilder } from "./PluginBuilder";

export interface SecretsManagerPluginObject {
  env?: Record<
    string, 
    string | {
      'secret-id': string;
      'json-key': string
    }
  >
}

type Env = Record<
  string, 
  string | {
    secretId: string;
    jsonKey: string
  }
>

export interface SecretsManagerPluginOptions {
}

export class SecretsManagerPlugin implements PluginBuilder {
  #env: Env | undefined;

  env(): Env | undefined;
  env(env: Env | undefined): this;
  env(env?: Env | undefined): Env | undefined | this {
    if (env) {
      this.#env = env
      return this
    } else {
      return this.#env;
    }
  }

  private toEnv(): SecretsManagerPluginObject['env'] {
    if (!this.#env) {
      return undefined;
    }
    const env: NonNullable<SecretsManagerPluginObject['env']> = {}
    for (const key in this.#env) {
      const value = this.#env[key]
      if (typeof value === 'string') {
        env[key] = value
      } else {
        env[key] = {
          "secret-id": value.secretId,
          "json-key": value.jsonKey,
        }
      }
    }
    return env
  }

  public build(): PluginObject<SecretsManagerPluginObject> {
    return {
      'seek-oss/aws-sm#v2.3.1': {
        env: this.toEnv()
      }
    }
  }
}