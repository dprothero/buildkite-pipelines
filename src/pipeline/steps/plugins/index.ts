export * from './PluginObject'
export * from './Plugin'
export * from './DockerPlugin'