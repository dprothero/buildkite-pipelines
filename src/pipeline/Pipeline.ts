import { PipelineObject } from "./PipelineObject";
import { StepBuilder } from "./steps/StepBuilder";
import { StepObject } from "./steps/StepObject";

function isBuilder(step: StepObject|StepBuilder): step is StepBuilder {
  return typeof (step as any).build == 'function'
}

export interface PipelineOptions {
  steps?: Array<StepObject | StepBuilder>
}

export class Pipeline {
  #steps: Array<StepObject | StepBuilder>;

  constructor(options: PipelineOptions) {
    this.#steps = options.steps ?? []
  }

  build(): PipelineObject {
    return {
      steps: this.#steps.map(step => {
        if (isBuilder(step)) {
          return step.build()
        } else {
          return step;
        }
      })
    }
  }
}