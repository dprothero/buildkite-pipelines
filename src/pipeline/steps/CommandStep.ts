import { CommandStepObject } from "./CommandStepObject";
import { PluginBuilder } from "./plugins/PluginBuilder";
import { buildKey, configureKey, KeyMixinOptions, mixinKey } from "./mixinKey";
import { StepBuilder } from "./StepBuilder";
import { buildLabel, configureLabel, LabelMixinOptions, mixinLabel } from "./mixinLabel";
import { buildDependency, configureDependency, DependenciesMixinOptions, mixinDependencies } from "./mixinDependencies";
import { BranchesMixinOptions, buildBranches, configureBranches, mixinBranches } from "./mixinBranches";
import { buildConditional, ConditionalMixinOptions, configureConditional, mixinConditional } from "./mixinConditional";
import { PluginObject } from "./plugins/PluginObject";

function isBuilder(step: PluginObject|PluginBuilder): step is PluginBuilder {
  return typeof (step as any).build == 'function'
}

export interface CommandStepOptions extends KeyMixinOptions, LabelMixinOptions, DependenciesMixinOptions, ConditionalMixinOptions, BranchesMixinOptions {
  command: string | string[]
  env?: Record<string, string>;
  plugins?: PluginBuilder[]
}

export class CommandStep extends mixinKey(mixinLabel(mixinDependencies(mixinConditional(mixinBranches(class {}))))) implements StepBuilder {
  #command: string | string[] = []
  #env: Record<string, string> = {};
  #plugins: Array<PluginObject | PluginBuilder> = [];
  #concurrency?: number;
  #concurrency_group?: string

  command(): string  | string[];
  command(command: string | string[]): this;
  command(command?: string | string[]): string | string[] | this {
    if (command) {
      this.#command = command
      return this
    } else {
      return this.#command;
    }
  }

  env(): Record<string, string>;
  env(env: Record<string, string>): this;
  env(env?: Record<string, string>): Record<string, string> | this {
    if (env) {
      this.#env = env
      return this
    } else {
      return this.#env;
    }
  }

  addEnv(name: string, value: string): this {
    this.#env[name] = value
    return this;
  }

  plugins(): Array<PluginObject | PluginBuilder>;
  plugins(plugins: Array<PluginObject | PluginBuilder>): this;
  plugins(plugins?: Array<PluginObject | PluginBuilder>): Array<PluginObject | PluginBuilder> | this {
    if (plugins) {
      this.#plugins = plugins
      return this
    } else {
      return this.#plugins;
    }
  }

  addPlugin(plugin: PluginObject | PluginBuilder): this {
    this.#plugins.push(plugin)
    return this;
  }

  withConcurrency(jobs: number, group: string) {
    this.#concurrency = jobs;
    this.#concurrency_group = group
    return this;
  }

  build(): CommandStepObject {
    const step: CommandStepObject = {
      ...buildKey(this),
      ...buildLabel(this),
      ...buildDependency(this),
      ...buildConditional(this),
      ...buildBranches(this),
    }

    if (Array.isArray(this.#command)) {
      if (this.#command.length > 1) {
        step.commands = this.#command
      } else if (this.#command.length === 1) {
        step.command = this.#command[0]
      } else {
        throw new Error('Invalid command')
      } 
    } else {
      step.command = this.#command
    }

    if (Object.keys(this.#env).length) {
      step.env = this.#env
    }

    if  (this.#plugins.length) {
      step.plugins = this.#plugins.map(plugin => isBuilder(plugin) ? plugin.build() : plugin)
    }
    
    return step
  }
}