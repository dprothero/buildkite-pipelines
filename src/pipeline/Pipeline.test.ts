import { stringify } from '../stringify'
import {Pipeline} from './Pipeline'
import { CommandStep, DockerPlugin } from './steps'
import fs from 'fs'
import { SecretsManagerPlugin } from './steps/plugins/SecretsManagerPlugin'
import { DockerLoginPlugin } from './steps/plugins/DockerLoginPlugin'

describe('Pipeline', () => {
  // test('build', () => {

  //   const p = new Pipeline({
  //     steps: [
  //       new CommandStep('yarn run lint'),

  //       new CommandStep([
  //           'cd dir',
  //           'yarn run build'
  //         ]),

  //     ]
  //   })
    
  //   expect(p.build()).toMatchSnapshot()
  // })


  // test('checks', () => {

  //   const typecheck = new CommandStep()
  //   .label(':typescript: check:workspaces (back-end)')
  //   .command([
  //     "set -e",
  //     "cd back-end",
  //     "yarn install --frozen-lockfile --development",
  //     "yarn run setup",
  //     "yarn run check:workspaces"
  //   ])
  //   .branches([ 'master', 'preview', 'production'])
  //   .plugins([new DockerPlugin()])
      
  //   const pipeline = new Pipeline({
  //     steps: [
  //       typecheck,

  //       new CommandStep([
  //         'cd dir',
  //         'yarn run build'
  //       ]),

  //     ]
  //   })

  //   expect(pipeline.build()).toMatchSnapshot() 
  // })


  test.only('secrets-docker-login-docker-cache', () => {
    const pipeline = new Pipeline().steps([
      new CommandStep()
        .label(':typescript: check:workspaces (back-end)')
        .command([
          'set -e',
          'cd back-end',
          'yarn install --frozen-lockfile --development',
          'yarn run setup',
          'yarn run check:workspaces'
        ])
        .branches([
          'master',
          'preview',
          'production'
        ])
        .plugins([
          new SecretsManagerPlugin().env({
            GITHUB_TOKEN: {
              secretId: 'global/buildkite/github',
              jsonKey: '.TOKEN'
            }
          }),
          new DockerLoginPlugin()
            .server('ghcr.io')
            .username('jameslnewell')
            .passwordEnv('GITHUB_TOKEN'),
          new DockerPlugin()
          .image('ghcr.io/mr-yum/ci:2.4.0')
        ])
    ])
    expect(stringify(pipeline.build())).toEqual(fs.readFileSync('./fixtures/secrets-docker-login-docker-cache.yml').toString())
  })
})

