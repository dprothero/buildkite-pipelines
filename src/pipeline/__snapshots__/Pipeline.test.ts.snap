// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pipeline build 1`] = `
Object {
  "steps": Array [
    Object {
      "command": "yarn run lint",
    },
    Object {
      "commands": Array [
        "cd dir",
        "yarn run build",
      ],
    },
  ],
}
`;

exports[`Pipeline checks 1`] = `
Object {
  "steps": Array [
    Object {
      "branches": Array [
        "master",
        "preview",
        "production",
      ],
      "commands": Array [
        "set -e",
        "cd back-end",
        "yarn install --frozen-lockfile --development",
        "yarn run setup",
        "yarn run check:workspaces",
      ],
      "label": ":typescript: check:workspaces (back-end)",
      "plugins": Array [
        Object {
          "docker#v3.10.0": Object {
            "always-pull": undefined,
            "environment": undefined,
            "image": undefined,
          },
        },
      ],
    },
    Object {
      "commands": Array [
        "cd dir",
        "yarn run build",
      ],
    },
  ],
}
`;
